profiler/export_profiles/main.go

```diff
@@ -0,0 +1,129 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// [START profiler_export_profiles]
+
+// Sample export_profiles shows how ListProfiles API can be used to download
+// existing pprof profiles for a given project from GCP.
+package main
+
+import (
+	"context"
+	"encoding/csv"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+
+	cloudprofiler "cloud.google.com/go/cloudprofiler/apiv2"
+	pb "cloud.google.com/go/cloudprofiler/apiv2/cloudprofilerpb"
+	"google.golang.org/api/iterator"
+)
+
+var project = flag.String("project", "", "GCP project ID from which profiles should be fetched")
+var pageSize = flag.Int("page_size", 100, "Number of profiles fetched per page. Maximum 1000.")
+var pageToken = flag.String("page_token", "", "PageToken from a previous ListProfiles call. If empty, the listing will start from the begnning. Invalid page tokens result in error.")
+var maxProfiles = flag.Int("max_profiles", 1000, "Maximum number of profiles to fetch across all pages. If this is <= 0, will fetch all available profiles")
+
+// This function reads profiles for a given project and stores them into locally created files.
+// The profile metadata gets stored into a 'metdata.csv' file, while the individual pprof files
+// are created per profile.
+func run(ctx context.Context) error {
+	client, err := cloudprofiler.NewExportClient(ctx)
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+	fmt.Printf("Attempting to fetch %v profiles with a pageSize of %v for %v\n", *maxProfiles, *pageSize, *project)
+
+	// Initial request for the ListProfiles API
+	request := &pb.ListProfilesRequest{
+		Parent:    fmt.Sprintf("projects/%s", *project),
+		PageSize:  int32(*pageSize),
+		PageToken: *pageToken,
+	}
+
+	// create a folder for storing profiles & metadata
+	if err := os.Mkdir("profiles", os.ModePerm); err != nil {
+		log.Fatal(err)
+	}
+	// create a file for storing profile metadata
+	metadata, err := os.Create("profiles/metadata.csv")
+	if err != nil {
+		return err
+	}
+	defer metadata.Close()
+
+	writer := csv.NewWriter(metadata)
+	defer writer.Flush()
+
+	writer.Write([]string{"File", "Name", "ProfileType", "Target", "Duration", "Labels"})
+
+	profileCount := 0
+	// Keep calling ListProfiles API till all profile pages are fetched or max pages reached
+	profilesIterator := client.ListProfiles(ctx, request)
+	for {
+		// Read individual profile - the client will automatically make API calls to fetch next pages
+		profile, err := profilesIterator.Next()
+
+		if err == iterator.Done {
+			log.Println("Read all available profiles")
+			break
+		}
+		if err != nil {
+			return fmt.Errorf("error reading profile from response: %v", err)
+		}
+		profileCount++
+
+		filename := fmt.Sprintf("profiles/profile%06d.pprof", profileCount)
```

Comment:
If you restarted the script with `-page_token=...` I think it would overwrite the previous files in the directory

Reply:
No, the script fails - because it tries to recreate an existing directory in the beginning. Seeing as it is just a sample showcasing one potential way to use the API, I feel ok with this.

Reply:
It's just a sample - you could remove the files from the location and re-run the script again with the last page token.

I could add additional handling in the code to check for existence of the folder - but I feel its not relevant to how ListProfiles API work.